### ISSUE TITLE
[Umbrella Ticket] PATH User Database 

### ISSUE THREAD

**Issue by Olshansk** [buildwithgrove/path#11]
## Objective

Build the foundation for tracking users accessing the portal.

## Origin Document

[Postgres driver](https://github.com/pokt-foundation/portal-middleware/pull/760)
[Proposed Database Structure](https://www.notion.so/buildwithgrove/Proposed-Database-Migration-8c6980560a474e038d272a3198467a35)

## Goals

- Associated a relay with a user ID
- Enable other gateways to build features associated with a specific user
- Enable downstream efforts related to billing
- Enable downstream efforts related to rate limiting
- Enable downstream efforts for user-specific features in the future (whitelisting, user agents, settings, etc...)

## Deliverables

- [ ] Implementation that ties a `relay` to a `UserId` including
	- [ ] Connecting to, fetching and caching user data from a Postgres database.
	- [ ] Relating a service request to this user data via a UserApp ID (eg from .../v1/{USER_APP_ID} in the URL).
- [ ] User-specific rate limiting (dependent on #10 as a parallel requirement)
- [ ] Updated tooling and documentation to startup a postgres DB alongside the gateway
- [ ] 1-2 pager w/ docs to be turned in 🦖  docs in the future

## Non-goals / Non-deliverables

- Out-of-the box stripe integration or hooks
- Implementing downstream user settings (whitelisting, user agents, settings, etc...)
- Integrating with crypto payments

## General deliverables

- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.
- [ ] **Testing**: Add new tests (unit and/or E2E) to the test suite.
- [ ] **Makefile**: Add new targets to the Makefile to make the new functionality easier to use.
- [ ] **Documentation**: Update architectural or development READMEs; use [mermaid](https://mermaid-js.github.io/mermaid/) diagrams where appropriate.

---

**Creator**: @Olshansk 
**Co-Owners**: @commoddity

**Comment by fredteumer** [comment]
this feels like it's being conflated from a user db vs billing db situation. The purpose of the database is to codify the business  logic of the application for the purposes of user management, plan management, services, etc.

Billing, while important, inherits a small subset of this data and is combined with the metering of the application to integrate with Stripe.

**Comment by commoddity** [comment]
> this feels like it's being conflated from a user db vs billing db situation. The purpose of the database is to codify the business logic of the application for the purposes of user management, plan management, services, etc.
> 
> Billing, while important, inherits a small subset of this data and is combined with the metering of the application to integrate with Stripe.

To be clear, I think the extent to which this issue needs to interact with billing is simply providing a `UserApp` ID we can relate to a relay to identify who sent the relay and, thus, who gets billed for it. 

I know we just discussed this a few mins ago but I tend to agree that adding ` Out-of-the box stripe integration or hooks` may be getting a bit ahead of ourselves for this iteration.

**Comment by fredteumer** [comment]
The above makes more sense for the scope of this ticket. We can expand upon it to open a metering/payment rail integration with stripe as a future feature and mark other possible payments integrations as future issues.

**Comment by Olshansk** [comment]
@commoddity @fredteumer PTAL at the updated goals, deliverables and non-goals and 👍 if it looks good to you.

I'm also putting together #13 to capture the billing work in a follow-up ticket.

**Comment by commoddity** [comment]
Draft PR for implementing Envoy Auth Plugin Proposal here:
https://github.com/buildwithgrove/path/pull/28
